### ISSUE TITLE
Estimated time for inference?

### ISSUE THREAD

**Issue by jamesdolezal** [adamshephard/OMTscoring_inference#1]
Hello - I was wondering if you are able to provide any reference benchmarks on the time required to generate whole-slide segmentations from the provided pretrained model. It's been running for about 2.5 hours (32-core CPU, A100s) on a single slide, and it's not really clear if it's making progress at all. I'm seeing:

```
Process Batch: 100%|############################| 25/25 [00:17<00:00,  1.40it/s]
Process Batch: 100%|############################| 25/25 [00:17<00:00,  1.41it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.35it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.36it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.35it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.35it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.37it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:17<00:00,  1.40it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.39it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.37it/s]
Process Batch: 100%|############################| 25/25 [00:18<00:00,  1.38it/s]
Process Batch: 100%|##############################| 4/4 [00:02<00:00,  1.55it/s]
...
```

... without a clear indication of the overall progress. I'm concerned that it might be stuck in an infinite loop somehow?

CPU utilization looks low, as does GPU utilization, but GPU utilization does spike which makes me think that it's doing some inference on GPU. 

For context, the slide is scanned at 20X, with dimensions around 45k x 40k. This should contain around 28,000 tiles at a size of 256x256.

A ballpark estimation of how long it's expected to take will be helpful to understand if something is wrong or if it's progressing as expected.

**Comment by adamshephard** [comment]
Hi @jamesdolezal,

Thanks for reaching out. The nuclei segmentation is done with HoVer-Net+, and is implemented with TIAToolbox. This generally enables quick and efficient inference on whole slide images. However, it does mean that there is a lack of overall progress bar, as you point out. The TIAToolbox team are currently aware about this issue and are planning to correct it in the future. Thank you for pointing this out again though. I will raise this in the next TIAToolbox meeting.

Unfortunately, we are unable to provide any benchmarks for this. On the CPU, HoVer-Net+ can be very slow. But for a slide of the dimensions you have specified, I wouldn't imagine that it would take more than an hour to process, especially with the A100 GPUs. Are you sure it is using the GPUs? TIAToolbox requires a single argument in the script to be True: https://github.com/adamshephard/OMTscoring_inference/blob/e9cf3d0d53868eaee505635ddb13b948fcb045e2/run_segmentation.py#L71

Alternatively, it may be best to provide a mask to the argument below. This may speed up processing further.
https://github.com/adamshephard/OMTscoring_inference/blob/e9cf3d0d53868eaee505635ddb13b948fcb045e2/run_segmentation.py#L68

Let me know how you get on! :-)

Best,
Adam

**Comment by jamesdolezal** [comment]
Thanks for the quick response! I did enable the GPU flag. I killed the process after 3 hours and tried again with:
- `num_loader_workers=32`
- `num_postproc_workers=32`
- `batch_size=128`

It seemed to work better this way - GPU utilization was overall higher. It did still take 2 hours and 40 minutes to generate for a single slide. However, it sounds like this is in the relative ballpark for what can be expected, which is helpful to know.

Thanks again!
